File: 1. Basic Topics/2. Backtracking/cpp/40. Combination Sum II.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> tmp;
        backtrack(candidates, target, ans, tmp, 0);
        return ans;
    }

    void backtrack(vector<int>& candidates, int target, vector<vector<int>>& ans, vector<int>& tmp, int start) {
        if(target == 0) {
            ans.push_back(tmp);
            return;
        }
        for(int i = start; i < candidates.size() && target >= candidates[i]; i++){
            if(i != start && candidates[i] == candidates[i - 1]){
                continue;
            } 
            tmp.push_back(candidates[i]);
            backtrack(candidates, target - candidates[i], ans, tmp, i + 1);
            tmp.pop_back();
        }
    }
};

File: 1. Basic Topics/2. Backtracking/py/40. Combination Sum II.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        ans = []
        tmp = []
        self.backtrack(candidates, target, ans, tmp, 0)
        return ans

    def backtrack(self, candidates: List[int], target: int, ans: List[List[int]], tmp: List[int], start: int) -> None:
        if target == 0:
            ans.append(list(tmp))
            return
        for i in range(start, len(candidates)):
            if i > start and candidates[i] == candidates[i - 1]:
                continue
            if target < candidates[i]:
                break
            tmp.append(candidates[i])
            self.backtrack(candidates, target - candidates[i], ans, tmp, i + 1)
            tmp.pop()